ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/TableExport.java

```diff
@@ -0,0 +1,271 @@
+    this.db = db;
+    this.conf = conf;
+    this.logger = logger;
+    this.paths = paths;
+  }
+
+  public AuthEntities run() throws SemanticException {
+    if (tableSpec == null) {
+      writeMetaData(null);
+    } else if (shouldExport()) {
+      if (tableSpec.tableHandle.isView()) {
+        replicationSpec.setIsMetadataOnly(true);
+      }
+      PartitionIterable withPartitions = partitions();
+      writeMetaData(withPartitions);
+      if (!replicationSpec.isMetadataOnly()) {
+        writeData(withPartitions);
+      }
+    }
+    return authEntities;
+  }
+
+  private PartitionIterable partitions() throws SemanticException {
+    try {
+      long currentEventId = db.getMSC().getCurrentNotificationEventId().getEventId();
+      replicationSpec.setCurrentReplicationState(String.valueOf(currentEventId));
+      if (tableSpec.tableHandle.isPartitioned()) {
+        if (tableSpec.specType == TableSpec.SpecType.TABLE_ONLY) {
+          // TABLE-ONLY, fetch partitions if regular export, don't if metadata-only
+          if (replicationSpec.isMetadataOnly()) {
+            return null;
+          } else {
+            return new PartitionIterable(db, tableSpec.tableHandle, null, conf.getIntVar(
+                HiveConf.ConfVars.METASTORE_BATCH_RETRIEVE_MAX));
+          }
+        } else {
+          // PARTITIONS specified - partitions inside tableSpec
+          return new PartitionIterable(tableSpec.partitions);
+        }
+      } else {
+        // Either tableHandle isn't partitioned => null, or repl-export after ts becomes null => null.
+        // or this is a noop-replication export, so we can skip looking at ptns.
+        return null;
+      }
+    } catch (Exception e) {
+      throw new SemanticException("Error when identifying partitions", e);
+    }
+  }
+
+  private void writeMetaData(PartitionIterable partitions) throws SemanticException {
+    try {
+      EximUtil.createExportDump(
+          paths.exportFileSystem,
+          paths.metaDataExportFile(),
+          tableSpec == null ? null : tableSpec.tableHandle,
+          partitions,
+          replicationSpec);
+      logger.debug("_metadata file written into " + paths.metaDataExportFile().toString());
+    } catch (Exception e) {
+      // the path used above should not be used on a second try as each dump request is written to a unique location.
+      // however if we want to keep the dump location clean we might want to delete the paths
+      throw new SemanticException(
+          ErrorMsg.IO_ERROR.getMsg("Exception while writing out the local file"), e);
+    }
+  }
+
+  private void writeData(PartitionIterable partitions) throws SemanticException {
+    try {
+      if (tableSpec.tableHandle.isPartitioned()) {
+        if (partitions == null) {
+          throw new IllegalStateException(
+              "partitions cannot be null for partitionTable :" + tableSpec.tableName);
+        }
+        for (Partition partition : partitions) {
```

Comment:
Can we avoid iterating through the partitions twice (once in writeMetaData and once here)? This will avoid lot of metastore calls and will improve dump performance significantly.